docs/src/debugging.md

```diff
@@ -72,14 +75,127 @@ ClimaCore.DebugOnly.call_post_op_callback() = false # hide
 ```
 This example should print `NaN` on your standard output.
 
-### Infiltrating
+As you see, this only tells us that a `NaN` was found, but not which function
+triggered the `NaN`. A simple way to find that is to extend this example with
+`Infiltrator`.
+
+#### Infiltrating and Exfiltrating
 
 [Infiltrator.jl](https://github.com/JuliaDebug/Infiltrator.jl) is a simple
 debugging tool for Julia packages.
 
 Here is an example, where we can use Infiltrator.jl to find where NaNs is coming
 from interactively.
 
+##### Infiltrating
+
+Suppose you have a `NaN` in your simulation and what to look at the variables
+right before the expression caused the `NaN`. One of the challenges in this is
+that you probably have a tower of function being called, many of which belong to
+other packages.
+
+Let us simulate this case
+```julia
+import ClimaCore
+using ClimaCore.CommonSpaces
+space = CubedSphereSpace(; radius = 10, n_quad_points = 4, h_elem = 10)
+myrho = ones(space)
+myP = ones(space)
+myu = ones(space)
+
+kinetic_energy(field) = field .* field ./ 2
+other_energy(field) = field ./ sum(field)
+offset_by_one(field) = field .- 1
+
+function specific_energy(rho, P, u)
+    density_without_restmass = offset_by_one(rho)
+    return (kinetic_energy(u) .+ other_energy(P)) ./ density_without_restmass
+end
+
+function renormalized_energy(rho, P, u)
+    energy = specific_energy(rho, P, u)
+    return energy ./ sum(energy)
+end
+
+any(isnan, renormalized_energy(myrho, myP, myu)) # true
+```
+
+To debug this, we first need to identify where the first `NaN` is produced. We
+use `DebugOnly.call_post_op_callback` and infiltrate.
+```julia
+import Infiltrator # must be in your default environment
+ClimaCore.DebugOnly.call_post_op_callback() = true
+function ClimaCore.DebugOnly.post_op_callback(result, args...; kwargs...)
+    has_nans = result isa Number ? isnan(result) : any(isnan, parent(result))
+    has_inf = result isa Number ? isinf(result) : any(isinf, parent(result))
+    @infiltrate has_nans || has_inf
+end
+```
+
+"Infiltrating" means being dropped into a new REPL where in the scope of the
```

Comment:
Is this a precise way to describe "infiltrating"?

Reply:
I just looked at the docs and they does seem to reflect this. The implementation seems a bit nebulous to me, though. I'm not exactly sure how this differs from `@exfiltrate`

Reply:
I find infiltrating simpler than exfiltrating because there's no safehouse and no scope pollution. It also stops the evaluation of the program at the infiltration point, but you can continue/toggle the infiltration point, and add as many infiltration points as you want.